### PR TITLE
*: After enabing local tso, when changing it to false, delete all dclocations

### LINKED ISSUE #5331: After enabling the parameter "enable-local-tso: true", if it is changed to false, the pd cluster will be unavailable
### What did you do?
After tested the `Local TSO` features, and close the `Local TSO` features

### What did you expect to see?
Expect normal, but report `global tso allocator maximum number` errors.

### Root Cause

Only change the  config from`enable-local-tso: true` to `enable-local-tso: false`. the allocator's metadata still persists in etcd. and the code:
https://github.com/tikv/pd/blob/a5c00c0c7ef3d933416c53638694f4457140fd9d/server/tso/global_allocator.go#L159-L165

will check `len(dcLocationMap) > 0`, we expect it be `==0`, then alloc tso normally.  and the `dcLocationMap` loads from:
https://github.com/tikv/pd/blob/a5c00c0c7ef3d933416c53638694f4457140fd9d/server/tso/allocator_manager.go#L646-L653

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by HuSharp**
Signed-off-by: husharp <jinhao.hu@pingcap.com>

### What problem does this PR solve?

Issue Number: Close #5331 

### What is changed and how does it work?

```commit-message
After enabing local tso, when changing it to false, delete all dclocations.
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/etcdutil/etcdutil.go [hunk 1]
```diff
@@ -129,6 +129,25 @@ func EtcdKVGet(c *clientv3.Client, key string, opts ...clientv3.OpOption) (*clie
 	return resp, nil
 }
 
+// EtcdKVDelete returns the etcd DeleteResponse by given key or key prefix
+func EtcdKVDelete(c *clientv3.Client, key string, opts ...clientv3.OpOption) (*clientv3.DeleteResponse, error) {
+	ctx, cancel := context.WithTimeout(c.Ctx(), DefaultRequestTimeout)
+	defer cancel()
```
Comment (JmPotato) [inline pkg/etcdutil/etcdutil.go:135]: I think it's more appropriate to use a context passed in.

```suggestion
func EtcdKVDelete(ctx context.Context, c *clientv3.Client, key string, opts ...clientv3.OpOption) (*clientv3.DeleteResponse, error) {
```

File: server/server.go [hunk 2]
```diff
@@ -396,6 +396,12 @@ func (s *Server) startServer(ctx context.Context) error {
 		func() time.Duration { return s.persistOptions.GetMaxResetTSGap() })
 	// Set up the Global TSO Allocator here, it will be initialized once the PD campaigns leader successfully.
 	s.tsoAllocatorManager.SetUpAllocator(ctx, tso.GlobalDCLocation, s.member.GetLeadership())
+	// When disable local TSO after enabling local TSO, we should clean up the local TSO allocator's dc location map and etcd.
```
Comment (JmPotato) [inline server/server.go:399]:
```suggestion
	// When disabled the Local TSO, we should clean up the Local TSO Allocator's meta info written in etcd if it exists.
```

File: server/server.go [hunk 3]
```diff
@@ -821,6 +827,11 @@ func (s *Server) GetServiceMiddlewareConfig() *config.ServiceMiddlewareConfig {
 	return cfg
 }
 
+// SetEnableLocalTSO sets enable-local-tso flag of Server .
```
Comment (JmPotato) [inline server/server.go:830]: Please add a comment to tell that this function should only be used in the test.

File: server/tso/allocator_manager.go [hunk 4]
```diff
@@ -245,6 +245,37 @@ func (am *AllocatorManager) GetDCLocationInfo(dcLocation string) (DCLocationInfo
...
 
+// CleanUpDCLocation cleans up certain server's DCLocationInfo
+func (am *AllocatorManager) CleanUpDCLocation() error {
+	am.mu.Lock()
+	defer am.mu.Unlock()
+
+	// Check if there is any dc-location to clean up
+	serverID := am.member.ID()
+	dcLocationKey := am.member.GetDCLocationPath(serverID)
+	if resp, err := kv.
```
Comment (JmPotato) [inline server/tso/allocator_manager.go:256]: Could we just try to delete it directly? I think a read before the deleting is unnecessary.

File: tests/server/tso/tso_test.go [hunk 5]
```diff
@@ -93,3 +94,55 @@ func requestLocalTSOs(re *require.Assertions, cluster *tests.TestCluster, dcLoca
...
+		conf.EnableLocalTSO = true
+		conf.Labels[config.ZoneLabel] = dcLocationConfig[serverName]
+	})
+	defer cluster.Destroy()
+	re.NoError(err)
+	re.NoError(cluster.RunInitialServers())
+
+	cluster.WaitAllLeaders(re, dcLocationConfig)
+	requestLocalTSOs(re, cluster, dcLocationConfig)
+	re.NoError(failpoint.Enable("github.com/tikv/pd/server/tso/systemTimeSlow", `return(true)`))
```
Comment (JmPotato) [inline tests/server/tso/tso_test.go:118]: Why enable this failpoint?

File: tests/server/tso/tso_test.go [hunk 6]
```diff
@@ -93,3 +94,55 @@ func requestLocalTSOs(re *require.Assertions, cluster *tests.TestCluster, dcLoca
...
+	}
+
+	ctx = grpcutil.BuildForwardContext(ctx, leaderServer.GetAddr())
+	tsoClient, err := grpcPDClient.Tso(ctx)
+	re.NoError(err)
+	defer tsoClient.CloseSend()
+	re.NoError(tsoClient.Send(req))
+	resp, err := tsoClient.Recv()
+	re.NoError(err)
+	re.NotNil(checkAndReturnTimestampResponse(re, req, resp))
```
Comment (JmPotato) [inline tests/server/tso/tso_test.go:147]: In addition to requesting a Global TSO, I think we can also call `GetClusterDCLocationsFromEtcd` to check if the number of existing DCs is as expected.

File: server/tso/allocator_manager.go [hunk 7]
```diff
@@ -245,6 +245,29 @@ func (am *AllocatorManager) GetDCLocationInfo(dcLocation string) (DCLocationInfo
 	return infoPtr.clone(), true
 }
 
+// CleanUpDCLocation cleans up certain server's DCLocationInfo
+func (am *AllocatorManager) CleanUpDCLocation() error {
+	am.mu.Lock()
+	defer am.mu.Unlock()
```
Comment (JmPotato) [inline server/tso/allocator_manager.go:251]: Why do we need the lock here? It seems that there is no access to the `am.mu` inside the function.

File: server/tso/allocator_manager.go [hunk 8]
```diff
@@ -1164,3 +1187,8 @@ func (am *AllocatorManager) nextLeaderKey(dcLocation string) string {
 func (am *AllocatorManager) EnableLocalTSO() bool {
 	return am.enableLocalTSO
 }
+
+// SetEnableLocalTSO set the value of AllocatorManager.enableLocalTSO.
+func (am *AllocatorManager) SetEnableLocalTSO(enableLocalTSO bool) {
```
Comment (JmPotato) [inline server/tso/allocator_manager.go:1192]: Why do we need this function? After `server.SetEnableLocalTSO(false)` and rebooting, `tso.NewAllocatorManager` will have the false value automaticlly.

File: server/tso/allocator_manager.go [hunk 9]
```diff
@@ -245,6 +245,26 @@ func (am *AllocatorManager) GetDCLocationInfo(dcLocation string) (DCLocationInfo
...
+		NewSlowLogTxn(am.member.Client()).
+		Then(clientv3.OpDelete(dcLocationKey)).
+		Commit(); err != nil {
+		return errs.ErrEtcdTxnInternal.Wrap(err).GenWithStackByCause()
+	} else if !resp.Succeeded {
+		return errs.ErrEtcdTxnConflict.FastGenByArgs()
+	}
+	log.Info("delete dc-location which in etcd",
+		zap.String("dcLocation-key", dcLocationKey),
+		zap.Uint64("server-id", serverID))
```
Comment (JmPotato) [inline server/tso/allocator_manager.go:263]:
```suggestion
	log.Info("try to delete the dc-location key previously written in etcd",
		zap.Uint64("server-id", serverID))
```

**Review by JmPotato — APPROVED** [review]

File: server/tso/allocator_manager.go [hunk 10]
```diff
@@ -245,6 +245,25 @@ func (am *AllocatorManager) GetDCLocationInfo(dcLocation string) (DCLocationInfo
...
+	// remove etcd
+	if resp, err := kv.
+		NewSlowLogTxn(am.member.Client()).
+		Then(clientv3.OpDelete(dcLocationKey)).
+		Commit(); err != nil {
+		return errs.ErrEtcdTxnInternal.Wrap(err).GenWithStackByCause()
+	} else if !resp.Succeeded {
+		return errs.ErrEtcdTxnConflict.FastGenByArgs()
+	}
+	log.Info("delete the dc-location key previously written in etcd",
```
Comment (rleungx) [inline server/tso/allocator_manager.go:261]: When there is no dc-location key existing in etcd, we don't need to print this log?

File: server/tso/allocator_manager.go [hunk 11]
```diff
@@ -245,6 +245,25 @@ func (am *AllocatorManager) GetDCLocationInfo(dcLocation string) (DCLocationInfo
 	return infoPtr.clone(), true
 }
 
+// CleanUpDCLocation cleans up certain server's DCLocationInfo
+func (am *AllocatorManager) CleanUpDCLocation() error {
+	serverID := am.member.ID()
+	dcLocationKey := am.member.GetDCLocationPath(serverID)
+	// remove etcd
```
Comment (rleungx) [inline server/tso/allocator_manager.go:252]:
```suggestion
	// remove dcLocationKey from etcd
```

File: server/tso/allocator_manager.go (hunk 10 above)
Comment (HuSharp) [reply server/tso/allocator_manager.go:261]: Do you think it's better to print the exact information after calling function GetClusterDCLocationsFromEtcd to make sure it doesn't exist, or to print an ambiguous log like "try to delete"?

**Review by nolouch — COMMENTED** [review]
overall lgtm，plz address the comments.

**Review by nolouch — APPROVED** [review]

**Pull request merged**